### ISSUE TITLE
The road to the native runtime free world

### ISSUE THREAD

**Issue by bkchr** [paritytech/substrate#10579]
We want to get rid off the native runtime. Initially the native runtime was added to overcome the performance "problems" of using wasmi for the runtime execution. However, with wasmtime we are around 2x slower than native. While this still isn't perfect, it is way waster than wasmi that is around 10x slower (all depending on the work load etc). While the performance plays an important role here there are also other things to consider, like ensuring that the native and wasm runtime are deterministic. We have already run into problems with [Polkadot](https://polkadot.network/blog/a-polkadot-postmortem-24-05-2021/) when it comes to this determinism problems. While we can ensure to use the exact same compiler version for building the Polkadot binary in our release process, we can not ensure that everyone else is also using the same compiler version as we used for the wasm build. 

There is also a nice write up and some discussions here: https://github.com/paritytech/substrate/issues/7288

To remove the native runtime, we need to fix some things before:

1. Make the wasm memory growable. We have already run into a situation of a OOM. This happened because we only allow a fixed amount of memory per wasm instance. We solved this problem temporarily by running the native runtime which can just use the entire system memory and then afterwards we released a new Polkadot node with an increased memory budget. So, we were lucky because we had the native runtime that saved our ass, but if we remove it it would not be there anymore. So, we should support growing of the wasm memory. It can still in the end allocate in maximum 4GB, but that should be way enough for all our use cases and combined with some proper monitoring we should be able to be informed before we hit this upper limit.
2. Remove runtime api requirement of statically checking that the runtime api is implemented. Currently when trying to use a runtime api, it is required to tell the compiler that a certain type is required that needs to implement a runtime api. This was done on purpose to give people better hints at compile time that they don't have implemented a certain api, but this only works by checking the native runtime (because we can not tell the compiler to check the wasm file). So, if we want to remove the native runtime we also need to remove this requirement on having the runtime api being implemented for a type. This can be solved by implementing the runtime apis for a different trait that will then be able to call into the runtime. We will also need to change some stuff, like checking that a runtime api is implemented on startup instead of failing later when the node is running.
3. Genesis is currently being created using the native runtime. We will also need to fix this. Maybe with an extra tool or whatever.

There are maybe still some other things to consider and to fix on the way, like only compiling the runtime for wasm instead of the current native & wasm build.

**Comment by burdges** [comment]
As an aside not strictly relevant here, we do envision having "runtime libraries" eventually, like SPREEs or perhaps large verifier keys ala halo2, which might be implemented in another way like a flat file of data, or if they undergo closer auditing then eBPF or native code loaded via dlopen, or a native generator for a verifier key.  Afaik all node should use identical deployment scheme, so we'd never walk back the consensus improvement described here.

**Comment by bkchr** [comment]
Okay, sounds interesting. I always thought that SPREE would be different wasm files that are just linked at runtime? 

Do we really want to native load code? Doesn't this destroy the entire advantage of using wasm? I mean then we could have done that from the beginning and could have distributed the runtime as a native library.

**Comment by burdges** [comment]
I've too little grasp about what "wasm files linked at runtime" means, but the SPREE would require process-like separation from the main PVF, and the SPREE invokes hostcalls that modify the SPREE state, for which the main PVF lacks permission.

At least in my mind that's impossible with "linked at runtime" so instead the SPREE exposes some crate linked at build time, but then itself creates new hostcalls or more SPREE-calls available to the whole PVF including its own part. 

Yes, SPREEs could be WASM but maybe 2x matters enough that a common enough SPREE would be native and well audited, or maybe doing those native builds limits our hardware choices too much?  I donno..

> Do we really want to native load code?

I donno, it's plausible we'd never want native code loaded, but much here remains unclear to me, like the performance hit eBPF takes.

I only raised the comment in case anything of this impacts how you remove the native runtime.  :)

**Comment by bkchr** [comment]
> I've too little grasp about what "wasm files linked at runtime" means, but the SPREE would require process-like separation from the main PVF, and the SPREE invokes hostcalls that modify the SPREE state, for which the main PVF lacks permission.

Yeah that makes sense and I did not thought about it, but this could just be a host call to communicate with SPREE and the SPREE could be anything. 

Aka we could have SPREE by default in wasm and could provide native compiled SPREE that ships with the node. If the requested SPREE version would not be supported by the node, we could just run the wasm version. However, I think whatever we do it is orthogonal to removing the native runtime. 


> I only raised the comment in case anything of this impacts how you remove the native runtime. :)

That is always a good idea nevertheless :)

**Comment by Wizdave97** [comment]
@bkchr  Would host functions be removed entirely along side the  native runtime? 
Or would there be a workaround for that?

**Comment by Garandor** [comment]
A question on this popped up at https://github.com/Manta-Network/Manta/pull/701

The gist of it is:
There already exists a commandline flag to force WASM execution: `--execution wasm`

While stopping validators on a nondeterminism issue is certainly bad and should be avoided ( e.g. using that same CLI flag ), it might be tolerable for other *full node* network participants - as a downtime on their functions ( serving RPC in our case ) is not nearly as critical to the network as a chain stall.

We might want to actually optimize non-critical nodes for the happy case where everything works and have the 2x speedup in runtime execution result in a higher number of RPCs served per node ( got no numbers on the actual performance impact yet, using the 2x number from @bkchr's initial description here, actual gain I guess will also depend on the weight used per block on average )

`native runtime free` as on the tin would remove this optionality of tolerating stability issues for performance gain, so IMHO it's not desirable, since you can already force the WASM runtime with the above CLI flag.

**Comment by bkchr** [comment]
Did you benchmark the execution? And do you have that many calls into the runtime on your RPC nodes?

**Comment by Garandor** [comment]
> Did you benchmark the execution? And do you have that many calls into the runtime on your RPC nodes?

No benchmarks yet. Can only intuit at this point that in the case of full-ish blocks and benchmark-scale machines a 2x performance penalty would hurt RPC throughput.

We do expect high RPC load on our nodes through custom wallet syncing logic in the near future.
I realize these RPC serving nodes can be easily scaled horizontally, but what I'm arguing here is that this change would condemn non-critical network participants to **waste resources** for a gain in stability which is not strictly necessary.

A proposal could be to force `--execution wasm` on validators but leave `--execution native_else_wasm` as default on full nodes, leaving them to opt into wasm execution with `--execution wasm`.

**Comment by tomaka** [comment]
A "2x" slowdown estimate in the runtime execution speed does in no way translate in a 2x slowdown in the number of RPC requests that can be served. The slowdown obviously depends on the kind of requests that you get, but I actually expect it to be basically 0. Until recently, PolkadotJS didn't even expose the possibility to call into the runtime through the RPC layer.

> got no numbers on the actual performance impact yet

I don't think it is reasonable argue that the performance hit is not worth the removal of the native runtime without having any number to back this up.

**Comment by Garandor** [comment]
> I don't think it is reasonable to argue that the performance hit is not worth the removal of the native runtime without having any number to back this up.

I agree, which is why I'm pointing to the general wasting of CPU cycles here. RPC is just our use-case. Other people will be running indexing (i.e. subsquid) or offchain-workers on their full nodes - they're all impacted by this change.

**Comment by tomaka** [comment]
If the runtime execution speed suddenly divided by 2, then the weights would be multiplied by 2 in return.

The CPU usage problem that you describe is indeed correct, but it's actually independent of the time it takes to execute. No matter whether it takes 5 microseconds or 5 milliseconds to verify for example a balance transfer, the weights are adjusted so that a node running on the reference machine will use 100% CPU if the chain is at maximum capacity.

**Comment by Garandor** [comment]
> No matter whether it takes 5 microseconds or 5 milliseconds to verify for example a balance transfer, the weights are adjusted so that a node running on the reference machine will use 100% CPU if the chain is at maximum capacity.

There's some subtlety here. Weights are always benchmarked with the WASM runtime, as it's obviously important for validators who always use it to not overflow execution time.

Full nodes executing the native runtime thus use weights created with the WASM runtime, but execute them 2x faster => lower CPU usage on full nodes executing native.

**Comment by wigy-opensource-developer** [comment]
> My basis for argument is the assumption that a node that's executing block weight close to the benchmark limit will have a high CPU load if it's exactly on spec. That would leave little free resources to handle RPCs into the runtime. If that same node now had ~50% idle CPU because of native execution finishing faster, it'd have plenty capacity for "extra" RPC work. Please correct me if this assumption is wrong

You also need to consider how much of the RPC requests are querying recent state and how much of them are interested in "archive" data. The current native approach only helps if most of the calls are querying recent data, and the node is upgraded soon after an on-chain runtime upgrade. But yeah, we need hard benchmarking data to see how much extra resources are needed to avoid this native runtime optimization.

**Comment by tomaka** [comment]
> You also need to consider how much of the RPC requests are querying recent state and how much of them are interested in "archive" data. The current native approach only helps if most of the calls are querying recent data,

There's no relationship between the runtime and querying data through RPC requests.
When an RPC requests wants to read a storage value, it just reads directly from the database. The runtime is completely unrelated to this.

**Comment by arkpar** [comment]
It is also worth mentioning that RPC throughput scales well with the number of CPU cores. Whilst block import is serialized at the moment. Weight calculations do not account for any extra cores, that may be used for serving RPC requests.

**Comment by kianenigma** [comment]
For the purpose of the master tutorial, what I would love to have is a fixed node that can ingest different runtimes (one at a time, of course), and run any of them. This should ideally be as simple as changing a single line in `Cargo.toml`, or simple be a `--runtime wasm.wasm`.

Will this be feasible once this is done? 

I think this is already feasible using the first ugly approach. Assuming we only do wasm execution, we just need a glue trait that would convey some types like `OpaqueBlock`, and a few other types needed for building the RPC extensions (`AccountId`, `Nonce`) and the GenesisConfig. 

As far as I know, genesis state will be built with wasm after this, and we are moving away from custom RPCs as well. So I think the amount of information that the client needs to know about the runtime is actually very very small. 

looking at `bin/minimal/node` is a good small example to demonstrate this.

**Comment by arkpar** [comment]
@kianenigma You can do that already with a custom chain spec. Generic substrate node should be able to sync any chain with substrate supported consensus. E.g. `cargo run --release --chain polkadot.json` in the substrate repo can sync polkadot.

**Comment by kianenigma** [comment]
> @kianenigma You can do that already with a custom chain spec. Generic substrate node should be able to sync any chain with substrate supported consensus. E.g. `cargo run --release --chain polkadot.json` in the substrate repo can sync polkadot.

That's a valuable new avenue that I had totally forgotten about, thanks! 

But -- in order to create a chain-spec, you still need some kind of fully-functioning client to be integrated with a wasm runtime. My ideal scenario is have a runtime that is itself capable of generating its initial state, and can be loaded into an omni-client that can run it with minimal configuration. 

of course, as long as the consensus in the runtime and client are matching. For the sake of simplicity, assume manual-seal is being used. 

For the tutorial, what I can do as the tutorial creator is to pre-generate all chain specs and let each step of the tutorial to have a `.json` file. 

This is much better, but I am still hoping that we can (optionally) reduce the dependency of the client on the runtime even further.

**Comment by bkchr** [comment]
> My ideal scenario is have a runtime that is itself capable of generating its initial state, an

This will be do-able after: https://github.com/paritytech/substrate/issues/13334

The wasm file will have an API to generate the genesis state/initial state.

> omni-client that can run it with minimal configuration.

This will not be possible after this pr. I mean yes we will move closer to this as the native runtime is gone. However, a lot of runtime api calls etc require that the node and the runtime know the type. This could also be changed, but would require to change a lot of things. (not impossible for sure)

But yeah, for your use case not that important.

**Comment by kianenigma** [comment]
> cargo run --release --chain polkadot.json

Actually this does not seem to work. The genesis config of the linked node's native runtime and the chain-spec need to have matching pallets for this approach to work. Otherwise I get an error that the provided chain spec has either missing pallets or too many of them.

Given this, I am somewhat baffled about why `cargo run --release -- --chain polkadot.json` does indeed work. 🤔

**Comment by kianenigma** [comment]
> cargo run --release --chain polkadot.json

Actually this does not seem to work. The genesis config of the linked node's native runtime and the chain-spec need to have matching pallets for this approach to work. Otherwise I get an error that the provided chain spec has either missing pallets or too many of them.

Given this, I am somewhat baffled about why `cargo run --release -- --chain polkadot.json` does indeed work. 🤔

**Comment by bkchr** [comment]
> The genesis config of the linked node's native runtime and the chain-spec need to have matching pallets for this approach to work.

The `polkadot.json` is a raw genesis config and this doesn't require the native runtime to load it. So, not sure what you have done there.

**Comment by kianenigma** [comment]
Then it must be because I have _not_ used `--raw` chain spec, and now I see what is the point of it! :))

**Comment by crystalin** [comment]
I don't know if I shared my opinion before but there are 2 important points to using Native runtime:
- Speed (running 1000 tests using native is significantly faster, even when using interpreted wasm (which doesn't spend time compiling at the boot)
- Debugging (So far, I wasn't able to get proper debugging of the wasm runtime (tried building with debug symbols and tell the wasmtime to debug it, but no success)

**Comment by kianenigma** [comment]
> This will not be possible after this pr. I mean yes we will move closer to this as the native runtime is gone. However, a lot of runtime api calls etc require that the node and the runtime know the type. This could also be changed, but would require to change a lot of things. (not impossible for sure)

I was just thinking about this again. I accept that it is not possible to decouple the two 100%, that's not what I am asking for. 

**What I am asking for to make the interface between the two clear.** 

As it stands now, the client of a substrate based chain is expected to arbitrarily imports the `runtime` crate to use its type. 

Moreover, the `runtime` code is assumed to know exactly which types it should declare as public in oder to satisfy a runtime. 

I think both of the above are bad situation. I don't know if it was left like this because it is not possible, or out of ignorance? The solution seems to be a trait to connect the two in a clear manner. 

```rust
/// The client's representation of a runtime. 
trait Runtime {
    type RuntimeApis; 
    type OpaqueBock; 
    type GenesisConfig; // This will not be needed anymore, but you get the point ;) 
}
```

**Comment by kianenigma** [comment]
> I would even go that far and say that you don't even need to import the runtime in your node anymore.

_Hell yeah_ is all I have to say to that then, yes none of what I said is needed.

**Comment by liamaharon** [comment]
Bumping @crystalin's comment here:

> I don't know if I shared my opinion before but there are 2 important points to using Native runtime:
> 
> * Speed (running 1000 tests using native is significantly faster, even when using interpreted wasm (which doesn't spend time compiling at the boot)
> * Debugging (So far, I wasn't able to get proper debugging of the wasm runtime (tried building with debug symbols and tell the wasmtime to debug it, but no success)

I've also found debugging of the wasm runtime very hit-or-miss, with lots of variables being logged as just empty strings. 

As try-runtime-cli develops and more debugging features are added, I'm wondering if there could be a benefit to using it with a native runtime.

**Comment by bkchr** [comment]
The native runtime can already not be used anymore. 

> with lots of variables being logged as just empty strings.

This is just because we have implemented a lot of `Debug` impls to do this. Generally this is something that we should revisit and to check the impact on the runtime. Generally, when things are not being used, the linker should strip them.

**Comment by athei** [comment]
> This is just because we have implemented a lot of `Debug` impls to do this. Generally this is something that we should revisit and to check the impact on the runtime. Generally, when things are not being used, the linker should strip them.

Log messages would probably the place which pull them into the runtime image. Solution would be to disable the lower log levels at compile time by default.

**Comment by athei** [comment]
I guess then it would be an interesting experiment to see by how much the size of a runtime goes up if we replace all `RuntimeDebug` by `Debug`. We could compare if there is a substantial difference when `sp-api/disable-logging` is set. With `lto` enabled the linker should be able to eliminate the unused impls.

**Comment by nazar-pc** [comment]
I tried once `RuntimeDebug` and noticed it was somehow more code with `RuntimeDebug` than `Debug` and I never used `RuntimeDebug` since. I didn't bother checking why it was the case though.
